### PR TITLE
 Enhance `label()` method in `ActiveField` with `tag` option and fix `labelOptions` for `checkbox`/`radio` (#20537).

### LINKED ISSUE #20537: ActiveField: Enhance `label()` method with `tag` option and fix `labelOptions` for `checkbox`/`radio`.
## What steps will reproduce the problem?

When using `ActiveField::checkbox()` or `ActiveField::radio()` methods with `$enclosedByLabel = false`  and providing a custom label string, the `labelOptions` array is completely ignored.

```php
$field->checkbox(
    [
        'label' => 'Custom Label',
        // This is ignored
        'labelOptions' => [
            'class' => 'custom-class',
            'data-test' => 'value',
            'tag' => 'span',
        ],
    ],
    false,
);
```

https://github.com/yiisoft/yii2/blob/de3bd9e592efedc929d5116d0e1c61052cdc9a12/framework/widgets/ActiveField.php#L614

##What is the expected result?

The labelOptions should be applied to the label element, allowing developers to:

- Add CSS classes and HTML attributes to the label.
- Change the label tag (for example, from `<label>` to  `<span>`).
- Use `tag` => `false` to render only the label text without any wrapper.

```php
public function testCheckboxEnclosedByLabelFalseWithCustomLabel(): void
{
    $this->activeField->checkbox(
        [
            'label' => 'Custom Label',
            'labelOptions' => [
                'class' => 'custom-label-class',
                'data-test' => 'custom-label-data',
            ],
        ],
        false,
    );

    $this->assertEqualsWithoutLE(
        <<<HTML
        <div class="form-group field-activefieldtestmodel-attributename">
        <label class="custom-label-class" data-test="custom-label-data" for="activefieldtestmodel-attributename">Custom Label</label>
        <input type="hidden" name="ActiveFieldTestModel[attributeName]" value="0"><input type="checkbox" id="activefieldtestmodel-attributename" name="ActiveFieldTestModel[attributeName]" value="1">
        <div class="hint-block">Hint for attributeName attribute</div>
        <div class="help-block"></div>
        </div>
        HTML,
        $this->activeField->render(),
        'Failed asserting that checkbox renders correctly.',
    );
}

public function testCheckboxEnclosedByLabelFalseWithCustomLabelTag(): void
{
    $this->activeField->checkbox(
        [
            'label' => 'Custom Label',
            'labelOptions' => [
                'class' => 'custom-label-class',
                'data-test' => 'custom-label-data',
                'tag' => 'span',
            ],
        ],
        false,
    );

    $this->assertEqualsWithoutLE(
        <<<HTML
        <div class="form-group field-activefieldtestmodel-attributename">
        <span class="custom-label-class" data-test="custom-label-data">Custom Label</span>
        <input type="hidden" name="ActiveFieldTestModel[attributeName]" value="0"><input type="checkbox" id="activefieldtestmodel-attributename" name="ActiveFieldTestModel[attributeName]" value="1">
        <div class="hint-block">Hint for attributeName attribute</div>
        <div class="help-block"></div>
        </div>
        HTML,
        $this->activeField->render(),
        'Failed asserting that checkbox renders correctly.',
    );
}

public function testCheckboxEnclosedByLabelFalseWithCustomLabelTagFalse(): void
{
    $this->activeField->checkbox(
        [
            'label' => 'Custom Label',
            'labelOptions' => [
                'tag' => false,
            ],
        ],
        false,
    );

    $this->assertEqualsWithoutLE(
        <<<HTML
        <div class="form-group field-activefieldtestmodel-attributename">
        Custom Label
        <input type="hidden" name="ActiveFieldTestModel[attributeName]" value="0"><input type="checkbox" id="activefieldtestmodel-attributename" name="ActiveFieldTestModel[attributeName]" value="1">
        <div class="hint-block">Hint for attributeName attribute</div>
        <div class="help-block"></div>
        </div>
        HTML,
        $this->activeField->render(),
        'Failed asserting that checkbox renders correctly.',
    );
}
```

### PR DESCRIPTION
**Pull request by terabytesoftw**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ❌

Previous Behavior:

- `label()` generates a standard `<label>` tag with the model attribute label.
```php
$field->label();
// output: <label class="control-label" for="model-attribute">Attribute Name</label>
```
- `label('Custom Text')` generates a `<label>` tag with custom text.
```php
$field->label('My Label');
// output: <label class="control-label" for="model-attribute">My Label</label>
```
- `label(false)` removes the label entirely from the field (empty string).
```php
$field->label(false);
// output: '' (empty string, no label rendered)
```
- `label($text, $options)` generates a label with custom options merged with `labelOptions`.
```php
$field->label('My Label', ['class' => 'custom-class']);
// output: <label class="custom-class" for="model-attribute">My Label</label>
```

New Behavior:

`tag` => `'label'` default uses `Html::activeLabel()` to generate a standard `<label>` element with the for attribute.
```php
$field->label(options: ['label' => 'My Label']);
// output: <label class="control-label" for="model-attribute">My Label</label>
```
`tag` => `false` renders the label content as raw HTML without any wrapper tag.
```php
$field->label(
    options: [
        'label' => 'My Label',
        'tag' => false,
    ],
);
// output: My Label
```
`tag` => `span`/`div`/`h3`/etc - uses `Html::tag()` to generate the specified element.
```php
$field->label(
    'Form Section',
    [
        'class' => 'section-title',
        'tag' => 'h3',
    ],
);
// output: <h3 class="section-title">Form Section</h3>

### CODE REVIEW

**Comment by terabytesoftw** [comment]
> maybe can reduce redundant code ? so it's more easy to maintain later

Yes it is a good idea.

**Comment by terabytesoftw** [comment]
> I'm thinking, if this way can support custom tag why dont make it consistent for
> 
> ```
> $field->checkbox( )->label('Custom Label', 'labelOptions' => [
>     'class' => 'radio-option-label',
>     'data-value' => 'choice-1',
>     'tag' => 'span',
> ]);
> ```

If it's a good idea, i'll do it in another PR.

**Comment by terabytesoftw** [comment]
> I'm thinking, if this way can support custom tag why dont make it consistent for
> 
> ```
> $field->checkbox( )->label('Custom Label', 'labelOptions' => [
>     'class' => 'radio-option-label',
>     'data-value' => 'choice-1',
>     'tag' => 'span',
> ]);
> ```

done.

**Review by samdark — APPROVED** [review]

**Review by Arhell — COMMENTED** [review]
lgtm

File: framework/widgets/ActiveField.php [hunk 1]
```diff
@@ -266,25 +266,42 @@ public function end()
...
+     *   - If set to `false`, the label content will be rendered as raw HTML without any wrapper tag.
+     *   - If set to `'label'`, uses [[Html::activeLabel()]] to generate a standard label element.
+     *   - If set to another tag name (e.g., `'span'`, `'div'`), uses [[Html::tag()]] to generate that element.
+     *
      * @return $this the field object itself.
      */
     public function label($label = null, $options = [])
     {
-        if ($label === false) {
+        if ($label === false || (isset($this->labelOptions['label']) && $this->labelOptions['label'] === false)) {
```
[the remark this answers is not in the document]
Comment (xicond) [reply framework/widgets/ActiveField.php]: I think user should allowed to overwrite the label with newer value, isn't it ?
```
If set to `false`, the label content will be rendered as raw HTML without any wrapper tag.
```
except it's null

[the remark this answers is not in the document]
Comment (terabytesoftw) [reply framework/widgets/ActiveField.php]: Correct the user can overwrite the label.

[testLabelPriorityOfContent](https://github.com/terabytesoftw/yii2/blob/821e0cf47809edcd567c310209444c97576cdf57/tests/framework/widgets/ActiveFieldTest.php#L246)

[the remark this answers is not in the document]
Comment (xicond) [reply framework/widgets/ActiveField.php]: I mean
```
$this->activeField(['label'=>false])->label('anything');
```
newer should be the final decision

File: tests/framework/widgets/ActiveFieldTest.php [hunk 2]
```diff
@@ -272,99 +440,121 @@ public static function hintDataProvider(): array
      * @param mixed $hint The hint content.
      * @param string $expectedHtml The expected HTML.
      */
-    public function testHint(mixed $hint, string $expectedHtml): void
+    public function testHint($hint, $expectedHtml): void
```
[the remark this answers is not in the document]
Comment (terabytesoftw) [reply tests/framework/widgets/ActiveFieldTest.php]: This would be BC changing the method signature.

File: framework/widgets/ActiveField.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (terabytesoftw) [reply framework/widgets/ActiveField.php]: Update docs.

File: framework/widgets/ActiveField.php [hunk 3]
```diff
@@ -264,27 +264,47 @@ public function end()
...
+     *   - If set to `false`, the label content will be rendered as raw HTML without any wrapper tag.
+     *   - If set to `'label'`, uses [[Html::activeLabel()]] to generate a standard label element.
+     *   - If set to another tag name (for example, `'span'`, `'div'`), uses [[Html::tag()]] to generate that element.
+     *
      * @return $this the field object itself.
      */
     public function label($label = null, $options = [])
     {
-        if ($label === false) {
+        if ($label === false || (isset($this->labelOptions['label']) && $this->labelOptions['label'] === false)) {
```
[the remark this answers is not in the document]
Comment (terabytesoftw) [reply framework/widgets/ActiveField.php]: done.

**Review by xicond — APPROVED** [review]

File: framework/widgets/ActiveField.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (terabytesoftw) [reply framework/widgets/ActiveField.php]: > I mean
> 
> ```
> $this->activeField(['label'=>false])->label('anything');
> ```
> 
> newer should be the final decision

@samdark How would you treat this edge case?

File: tests/framework/widgets/ActiveFieldTest.php [hunk 4]
```diff
@@ -158,103 +167,301 @@ public function testBegin(): void
...
+        {
+            $this->activeField->labelOptions = ['label' => false];
+
+            $this->activeField->label();
+
+            $this->assertEmpty(
+                $this->activeField->parts['{label}'],
+        'Failed asserting that label does not render.',
+            );
+        }
```
[the remark this answers is not in the document]
Comment (terabytesoftw) [reply tests/framework/widgets/ActiveFieldTest.php]: fixed.

File: tests/framework/widgets/ActiveFieldTest.php [hunk 5]
```diff
@@ -158,103 +167,301 @@ public function testBegin(): void
...
+
+        public function testLabelOptionsLabelFalseWithNoParameters(): void
+        {
+            $this->activeField->labelOptions = ['label' => false];
+
+            $this->activeField->label();
+
+            $this->assertEmpty(
+                $this->activeField->parts['{label}'],
+        'Failed asserting that label does not render.',
```
[the remark this answers is not in the document]
Comment (terabytesoftw) [reply tests/framework/widgets/ActiveFieldTest.php]: fixed.

File: tests/framework/widgets/ActiveFieldTest.php [hunk 6]
```diff
@@ -272,99 +479,121 @@ public static function hintDataProvider(): array
      * @param mixed $hint The hint content.
      * @param string $expectedHtml The expected HTML.
      */
-    public function testHint(mixed $hint, string $expectedHtml): void
+    public function testHint($hint, $expectedHtml): void
```
[the remark this answers is not in the document]
Comment (terabytesoftw) [reply tests/framework/widgets/ActiveFieldTest.php]: done.

File: framework/widgets/ActiveField.php [hunk 7]
```diff
@@ -276,15 +286,35 @@ public function label($label = null, $options = [])
...
-            $options['for'] = null;
+        if ($label === null) {
+            $label = $options['label'] ?? null;
+
+            if ($label === false) {
+                $this->parts['{label}'] = '';
+
+                return $this;
+            }
         }
```
[the remark this answers is not in the document]
Comment (terabytesoftw) [reply framework/widgets/ActiveField.php:299]: https://github.com/terabytesoftw/yii2/blob/6d9f1c6c188b84251a122db38c793391553699f1/framework/widgets/ActiveField.php#L270

File: framework/widgets/ActiveField.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (terabytesoftw) [reply framework/widgets/ActiveField.php]: Label precedence.

- `$label` parameter (highest priority).
- `options['label']` (only when `$label` === `null`).
- `Model attribute label (fallback)`.
- If any of these equals `false`, the label won't render.

File: tests/framework/widgets/ActiveFieldTest.php [hunk 8]
```diff
@@ -1,10 +1,13 @@
 <?php
+
 /**
  * @link https://www.yiiframework.com/
  * @copyright Copyright (c) 2008 Yii Software LLC
  * @license https://www.yiiframework.com/license/
  */
 
+declare(strict_types=1);
+
```
[the remark this answers is not in the document]
Comment (terabytesoftw) [reply tests/framework/widgets/ActiveFieldTest.php:10]: The tests pass correctly, ensuring strict types.

File: tests/framework/widgets/ActiveFieldTest.php [hunk 9]
```diff
@@ -84,72 +87,78 @@ public function testRenderWithCallableContent(): void
...
-<input type="text" id="activefieldtestmodel-attributename" class="form-control" name="ActiveFieldTestModel[{$this->attributeName}]">
-<div class="hint-block">Hint for attributeName attribute</div>
-<div class="help-block"></div>
-</div> </div>
-</div>
-EOD;
-
-        $actualValue = $this->activeField->render($content);
-        $this->assertEqualsWithoutLE($expectedValue, $actualValue);
+        $content = static fn (string $field) => "<div class=\"custom-container\"> $field </div>";
```
[the remark this answers is not in the document]
Comment (terabytesoftw) [reply tests/framework/widgets/ActiveFieldTest.php]: Branch `22` uses PHP `8.1`, and in the tests, we use strict typing.

File: tests/framework/widgets/ActiveFieldTest.php [hunk 10]
```diff
@@ -674,7 +914,7 @@ public function testInputOptionsTransferToWidget(): void
         $this->assertStringContainsString('placeholder="pholder_direct"', (string) $widget);
 
         // use regex clientOptions instead mask
-        $widget = $this->activeField->widget(TestMaskedInput::className(), [
+        $widget = $this->activeField->widget(TestMaskedInput::class, [
```
[the remark this answers is not in the document]
Comment (terabytesoftw) [reply tests/framework/widgets/ActiveFieldTest.php:917]: The `className()` method is deprecated, `class` is preferred.

File: framework/widgets/ActiveField.php [hunk 11]
```diff
@@ -276,15 +286,35 @@ public function label($label = null, $options = [])
...
 
-        $this->parts['{label}'] = Html::activeLabel($this->model, $this->attribute, $options);
+        $label ??= $this->model->getAttributeLabel(Html::getAttributeName($this->attribute));
```
[the remark this answers is not in the document]
Comment (terabytesoftw) [reply framework/widgets/ActiveField.php:301]: The minimum version of the `22` branch is PHP `8.1`, so that's fine.

File: framework/widgets/ActiveField.php [hunk 12]
```diff
@@ -961,4 +988,46 @@ protected function addErrorClassIfNeeded(&$options)
...
+                $this->labelOptions = array_merge($this->labelOptions, $options['labelOptions']);
+            }
+
+            if ($tag === false) {
+                $this->parts['{label}'] = $options['label'];
+            } elseif ($tag === 'label') {
+                $this->parts['{label}'] = Html::activeLabel(
+                    $this->model,
+                    $this->attribute,
+                    ['label' => $options['label']] + $this->labelOptions,
```
[the remark this answers is not in the document]
Comment (terabytesoftw) [reply framework/widgets/ActiveField.php]: done.

File: framework/widgets/ActiveField.php [hunk 13]
```diff
@@ -961,4 +988,46 @@ protected function addErrorClassIfNeeded(&$options)
...
+    /**
+     * Generates the label element based on `labelOptions` configuration when `enclosedByLabel` is `false`.
+     *
+     * @param array $options the input options array
+     *
+     * @return array the cleaned options array with `labelOptions` removed
+     */
+    protected function generateLabel(array $options): array
+    {
+        if (isset($options['label']) && $this->parts['{label}'] === '') {
```
[the remark this answers is not in the document]
Comment (terabytesoftw) [reply framework/widgets/ActiveField.php]: done.

File: framework/widgets/ActiveField.php (hunk 11 above)
[the remark this answers is not in the document]
Comment (terabytesoftw) [reply framework/widgets/ActiveField.php:301]: Version `2.2` supports php `8.1`, which is a major version.

File: tests/framework/widgets/ActiveFieldTest.php [hunk 14]
```diff
@@ -158,103 +167,301 @@ public function testBegin(): void
...
+
+    public function testLabelWithLabelOptionsAndTagFalse(): void
+    {
+        $label = 'Label Name';
+
+        $this->activeField->label(
+            $label,
+            [
+                'class' => 'custom-class',
+                'tag' =>  false,
```
[the remark this answers is not in the document]
Comment (terabytesoftw) [reply tests/framework/widgets/ActiveFieldTest.php]: fixed.

File: framework/widgets/ActiveField.php [hunk 15]
```diff
@@ -961,4 +993,46 @@ protected function addErrorClassIfNeeded(&$options)
...
+            } elseif ($tag === 'label') {
+                $this->parts['{label}'] = Html::activeLabel(
+                    $this->model,
+                    $this->attribute,
+                    array_merge(['label' => $options['label']], $this->labelOptions),
+                );
+            } else {
+                $tagOptions = $this->labelOptions;
+
+                $this->parts['{label}'] = Html::tag($tag, $options['label'], $tagOptions);
```
[the remark this answers is not in the document]
Comment (terabytesoftw) [reply framework/widgets/ActiveField.php]: fixed.

File: tests/framework/widgets/ActiveFieldTest.php [hunk 16]
```diff
@@ -158,103 +167,301 @@ public function testBegin(): void
...
-        $expectedValue = '<label class="control-label" for="activefieldtestmodel-attributename">Attribute Name</label>';
         $this->activeField->label();
 
-        $this->assertEquals($expectedValue, $this->activeField->parts['{label}']);
+        $this->assertEqualsWithoutLE(
+            <<<HTML
+            <label class="control-label" for="activefieldtestmodel-attributename">Attribute Name</label>
+            HTML,
+           $this->activeField->parts['{label}'],
+       );
```
[the remark this answers is not in the document]
Comment (terabytesoftw) [reply tests/framework/widgets/ActiveFieldTest.php:434]: fixed.
